### PR TITLE
End to end prototype

### PR DESCRIPTION
**Pull request by KinanBab**
End to end prototype with just enough features to handle the following sample P4 program:
https://github.com/googleinterns/p4-symbolic/blob/end2end/p4-samples/port-table/table.p4

The workflow proceeds as follows:
1. p4c is run on the p4 program to produce bmv2 json and p4info (use `` to do this)
1. Bmv2 JSON is parsed into a primitive protobuf that matches the bmv2 exactly.
2. p4info is parsed with pdpi
3. bmv2 protobuf, pdpi, and table entries are combined and transformed into our IR representation
4. our IR representation is fed to the Analyzer class, which traverses the program, creating a bunch of symbolic variables and constraints on them guided by the IR.
5. The analyzer class can be queried by giving it a table name and a row in that table, and it will generate a packet that matches that row. This can be repeated many times.

The prototype can be run using:
```bash
# from repo root
# produce .json and p4info files
cd p4-samples/port-table && make build
# compile our project
bazel build //p4_symbolic:main
# run the binary on the sample program
./bazel-bin/p4_symbolic/main --p4info=p4-samples/port-table/table.pb.txt --bmv2=p4-samples/port-table/table.json --entries=p4-samples/port-table/entries.txt
```

The binary will output a debugging smt program, followed by the testing packets.

Remaining tasks to complete this PR: 
- [x] Refactor p4_symbolic/main.cc into an IR testing file and a general binary.
- [x] Extract packet in a better format from model and write to a file.
- [x] Dump debugging smt program to a file, if flagged.
- [x] Minor code clean ups (AnalyzeTable function is too big).
- [x] Add generic functions for iterating over tables and rows, and use them in the binary, instead of using hardcoded values.

Shortcuts and unsupported features are marked clearly with TODOs in the respective location of the code. These will be addressed in future PRs.

### CODE REVIEW

**Comment by KinanBab** [comment]
Test refactoring is done, but I pushed it to a separate branch 'refactor-testing' to keep this PR small.

File: p4_symbolic/main.cc [hunk 1]
```diff
@@ -106,7 +108,39 @@ int main(int argc, char *argv[]) {
     return 1;
   }
 
-  std::cout << ir_status.value().DebugString() << std::endl;
+  // Analyze program symbolically.
+  p4_symbolic::symbolic::Analyzer analyzer;
+  p4_symbolic::ir::P4Program program = ir_status.value();
```
Comment (konne88) [inline p4_symbolic/main.cc]: You should use RETURN_IF_ERROR/ASSIGN_OR_RETURN/RET_CHECK macros for error handling, so avoid distracting from the main code. Here, I would recommend pulling all the code into a Main function that returns a Status, and then do the std::cerr << status; return 1; only once.

File: p4_symbolic/symbolic/symbolic.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "absl/status/status.h"
+#include "p4_pdpi/utils/status_utils.h"
+#include "p4_symbolic/ir/ir.pb.h"
+#include "z3++.h"  // added as a system dependency for now.
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.h]: can you put

TODO(your google username): added as ...

File: p4_symbolic/symbolic/symbolic.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  // entries_map['table1_3'] <=> row at index 3 in table1 was hit.
+  std::unordered_map<std::string, z3::expr> kEntriesMap;
+
+  // Maps an action full name to the corresponding symbolic expressions
+  // representing that this action was indeed matched on and called.
+  std::unordered_map<std::string, z3::expr> kActionsMap;
+
+  // Symbolic constraints that describe relationships between various variables
+  // and fields in the program.
+  std::vector<z3::expr> kConstraints;
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.h]: Are these constants? If so, define them const, if not, they should be lowercase and end with _

File: p4_symbolic/symbolic/symbolic.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  // Class is not copyable or movable.
+  Analyzer(const Analyzer&) = delete;
+  Analyzer& operator=(const Analyzer&) = delete;
+  Analyzer(Analyzer&&) = delete;
+  Analyzer& operator=(Analyzer&&) = delete;
+
+  // Entry point
+  absl::Status Analyze(ir::P4Program);
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.h]: Please use explicit dataflow. It's very hard to tell what is going on here, when every function you are calling can mutate the entire state. I propose you remove this class, and have functions that take and return only the data they need.

File: p4_symbolic/symbolic/symbolic.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  // Dumps the SMT program for debugging.
+  std::string DebugSMT();
+
+  // API for finding test packets after analysis.
+  pdpi::StatusOr<std::map<std::string, std::string>> FindPacketHittingRow(
+      const std::string&, int);
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.h]: give all function parameters a name, everywhere

File: p4_symbolic/symbolic/symbolic.h (hunk 3 above)
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.h]: Instead of creating constraints, I think we should write the code to create symbolic expressions, and only at the very end do we assert something. In particular, the entire switch pipeline should be a symbolic expression. The expression should have the access to the input packet, and should produce a trace with all the information about how the packet was forwarded (which table entry was hit, was the packet dropped, etc). You can look at the code that I shared with you for more detail. The user can then write an assertion over that expression, e.g. packet must not have dropped, and must have hit table entry FOO. 

If you follow that style, your program will look very nice, because it looks just like a non-symbolic P4 simulator that happens to operate with symbolic values and operations.

**Review by konne88 — CHANGES_REQUESTED** [review]

File: p4_symbolic/symbolic/symbolic.h (hunk 3 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/symbolic.h]: yes this is my bad, I misread the style guide

File: p4_symbolic/main.cc (hunk 1 above)
Comment (konne88) [reply p4_symbolic/main.cc]: Why did you mark this resolved? It doesn't look changed to me

File: p4_symbolic/symbolic/symbolic.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+namespace p4_symbolic {
+namespace symbolic {
+
+// Maps packet field name to its concrete value.
+// e.g. "metadata.ingress_port" to 1.
+using Packet = std::map<std::string, std::string>;
+
+// Symbolically Analyzes the given program, generating z3 symbolic expressions
+// and that are later solved to produce concerete packets.
+pdpi::StatusOr<SolverState> AnalyzeProgram(const ir::P4Program &program);
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.h]: symbolic_ingress_port = FreePort();
symbolic_ingress_header = FreeHeader();
symbolic_trace = RunP4PipelineSymbolically(p4_program, table_entries, symbolic_ingress_port, symbolic_ingress_header))

File: p4_symbolic/symbolic/structs.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  ir::P4Program program;
+  // Z3 context wrapped in a smart pointer.
+  // z3::context implements a private copy constructor and no move contructor.
+  // This makes the call tostd::move(<context>) in pdpi::StatusOr
+  // fail with a compile time error.
+  // Z3's context is not meant to be copied or moved, related symbolic
+  // expressions must be created using the exact same context.
+  // Using a smart pointer address these issues and automatically manages the
+  // memory.
+  std::unique_ptr<z3::context> context;
```
Comment (konne88) [inline p4_symbolic/symbolic/structs.h]: Don't use this inside of P4 pipeline, create all variables/assertions outside.

File: p4_symbolic/symbolic/table.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+#include "p4_pdpi/utils/status_utils.h"
+#include "p4_symbolic/ir/ir.pb.h"
+#include "p4_symbolic/symbolic/structs.h"
+#include "z3++.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+namespace table {
+
+pdpi::StatusOr<z3::expr> AnalyzeTable(const ir::Table &table,
```
Comment (konne88) [inline p4_symbolic/symbolic/table.h]: trace = RunTable(p4_table, table_entries, state)

trace consists of new state, and then table entry that was hit to give to the assertion

File: p4_symbolic/symbolic/action.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+// variables and symbolic values in its scope.
+struct ActionContext {
+  std::string action_name;
+  // Variable name to its corresponding symbolic expression.
+  std::unordered_map<std::string, z3::expr> scope;
+};
+
+// Symbolically Analyzes the given program, generating z3 symbolic
+// expressions and that are later solved to produce concerete packets.
+pdpi::StatusOr<ActionSymbolicTrace> AnalyzeAction(
```
Comment (konne88) [inline p4_symbolic/symbolic/action.h]: state = RunAction(action, state)

Comment (konne88) [reply p4_symbolic/symbolic/action.h]: Rerun action analysis every time.

File: p4_symbolic/symbolic/names.h [hunk 10]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 Google LLC
```
Comment (konne88) [inline p4_symbolic/symbolic/names.h]: This will go away

File: p4_symbolic/symbolic/structs.h [hunk 11]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 Google LLC
```
Comment (konne88) [inline p4_symbolic/symbolic/structs.h]: Put these with their appropriate functions

**Review by konne88 — CHANGES_REQUESTED** [review]

File: p4_symbolic/symbolic/symbolic.cc [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+#include "p4_symbolic/symbolic/symbolic.h"
+
+#include "p4_symbolic/symbolic/table.h"
+#include "p4_symbolic/symbolic/util.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+
+z3::context *Z3_CONTEXT = nullptr;
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.cc]: Can we just initialize this variable once, and keep it forever? How the code initializes and frees it right now, is hard to follow for me.

File: p4_symbolic/symbolic/symbolic.cc [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+  // Construct solver state for this program.
+  SolverState *solver_state =
+      new SolverState{data_plane.program, data_plane.entries, symbolic_context,
+                      std::unique_ptr<z3::context>(Z3_CONTEXT),
+                      std::unique_ptr<z3::solver>(z3_solver)};
+
+  return solver_state;
+}
+
+pdpi::StatusOr<ConcreteContext> Solve(SolverState *solver_state,
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.cc]: Can we change this to optional? It's not really an error if the solver doesn't find a solution.

File: p4_symbolic/symbolic/symbolic.cc [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+#include "p4_symbolic/symbolic/table.h"
+#include "p4_symbolic/symbolic/util.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+
+z3::context *Z3_CONTEXT = nullptr;
+
+pdpi::StatusOr<SolverState *> EvaluateP4Pipeline(
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.cc]: Is there are reason to return the pointer? Why not just return the value?

File: p4_symbolic/symbolic/symbolic.cc [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    const Dataplane &data_plane, const std::vector<int> &physical_ports) {
+  // Check Z3_CONTEXT is freed properly to avoid any context sharing issues.
+  if (Z3_CONTEXT != nullptr) {
+    return absl::Status(absl::StatusCode::kAlreadyExists,
+                        "Cannot make subsequent call to EvaluateP4Pipeline() "
+                        "when SolverState returned by a previous call has not "
+                        "been destructed!");
+  }
+
+  // Context to use for defining z3 variables, etc..
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.cc]: Instead of initializing the context here, can we just initialize it once for the whole program. There is a lot of subtly around global variables that have destructors (go/totw/110), but the TLDR is:

z3::context& Z3Context() {
  static z3::context* context = new z3::context();
  return *context;
}

File: p4_symbolic/symbolic/symbolic.cc [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  Z3_CONTEXT = new z3::context();
+  z3::solver *z3_solver = new z3::solver(*Z3_CONTEXT);
+
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.cc:35]: You should basically never call new. Can you create a unique_ptr instead?

**Review by konne88 — CHANGES_REQUESTED** [review]
Looks great. My main feedback is that I think the global variable initialization and destruction seems too complex. Can we just initialize the context once on program start?

File: p4_symbolic/symbolic/symbolic.cc (hunk 13 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/symbolic.cc]: Makes sense

File: p4_symbolic/symbolic/symbolic.cc (hunk 14 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/symbolic.cc]: Yes, the SolverState has a z3::context and z3::solver in it. These do not have copy constructors. pdpi::StatusOr<> requires a copy constructor. Also, because I wanted SolverState to be essentially a singleton.

**Comment by KinanBab** [comment]
Makes sense. It is a bit annoying how the context is managed right now. I can make this change to simplify it.

File: p4_symbolic/symbolic/symbolic.cc (hunk 15 above)
Comment (KinanBab) [reply p4_symbolic/symbolic/symbolic.cc]: Actually, there is no need for Z3_CONTEXT to even be a pointer.

Comment (konne88) [reply p4_symbolic/symbolic/symbolic.cc]: So if you have the Z3Context function, you just always access the context through that function. Because the context inside the function is static, it's essentially a global variable. Inside the function, the context has to be a pointer, otherwise, the object would be destructed when the function returns (I believe).

Comment (konne88) [reply p4_symbolic/symbolic/symbolic.cc]: Oh actually, the destructor will be called at the end of the program, but I believe we don't want that, due to some subtle race conditions etc as described in go/totw/110

**Comment by KinanBab** [comment]
Your comments where very helpful. I did some more digging, and got rid of all pointers. Z3_CONTEXT is now just a plain value that is always initialized. SolverState is returned as a unique_ptr.

File: p4_symbolic/symbolic/symbolic.cc [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+#include <utility>
+
+#include "p4_symbolic/symbolic/table.h"
+#include "p4_symbolic/symbolic/util.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+
+z3::context &Z3_CONTEXT() {
```
Comment (konne88) [inline p4_symbolic/symbolic/symbolic.cc]: Wrong naming style for a function. Call it Z3Context.

**Review by konne88 — APPROVED** [review]
